FlowCrypt/src/main/java/com/flowcrypt/email/security/pgp/PgpMsg.kt

```diff
@@ -65,4 +81,198 @@ object PgpMsg {
+    }
+
+    val chunk = data.copyOfRange(0, data.size.coerceAtMost(50)).toString(StandardCharsets.US_ASCII)
+    var input = data.inputStream() as InputStream
+
+    if (chunk.contains(ARMOR_HEADER_DICT[MsgBlock.Type.SIGNED_MSG]!!.begin)) {
+      val msg = PgpArmor.readSignedClearTextMessage(input)
+      return DecryptionResult.withCleartext(msg.content, msg.signature)
+    }
+
+    val isArmored = chunk.contains(ARMOR_HEADER_DICT[MsgBlock.Type.ENCRYPTED_MSG]!!.begin)
+    if (isArmored) input = ArmoredInputStream(input)
+
+    val protector = if (passphrase != null) {
+      PasswordBasedSecretKeyRingProtector.forKey(pgpSecretKeyRing, passphrase)
+    } else {
+      UnprotectedKeysProtector()
+    }
+
+    try {
+      val d = PGPainless.decryptAndOrVerify()
+          .onInputStream(input)
+          .decryptWith(protector, PGPSecretKeyRingCollection(listOf(pgpSecretKeyRing)))
+
+      val decryptionStream = if (pgpPublicKeyRingCollection != null) {
+        d.verifyWith(pgpPublicKeyRingCollection).ignoreMissingPublicKeys().build()
+      } else {
+        d.doNotVerify().build()
+      }
+
+      val output = ByteArrayOutputStream()
+      decryptionStream.use { it.copyTo(output) }
+
+      val streamResult = decryptionStream.result
+      return DecryptionResult.withDecrypted(output, streamResult.fileInfo?.fileName)
+    } catch (ex: MessageNotIntegrityProtectedException) {
+      return DecryptionResult(
+        content = null,
+        isEncrypted = true,
+        filename = null,
+        signature = null,
+        error = DecryptionError(
+          DecryptionErrorType.NO_MDC,
+          "Security threat! Message is missing integrity checks (MDC). The sender should " +
+          "update their outdated software."
+        )
+      )
+    } catch (ex: ModificationDetectionException) {
+      return DecryptionResult(
+          content = null,
+          isEncrypted = true,
+          filename = null,
+          signature = null,
+          error = DecryptionError(
+              DecryptionErrorType.BAD_MDC,
+              "Security threat! Integrity check failed."
+          )
+      )
+    } catch (ex: Exception) {
+      if (ex is PGPDataValidationException || (ex is PGPException
+              && (
+                ex.message?.contains("exception decrypting session info") == true
+                || ex.message?.contains("encoded length out of range") == true
+                || ex.message?.contains("Invalid Curve25519 public key") == true
+                || ex.message?.contains("Exception recovering session info") == true
+                || ex.message?.contains("ECDH requires use of PGPPrivateKey for decryption") == true
+              )
+          )
+      ) {
+        return if (passphrase != null) {
+          DecryptionResult.withError(
+              type = DecryptionErrorType.WRONG_PASSWORD,
+              message = "Wrong password",
+              cause = ex
+          )
```

Comment:
This seems a naive way to do it - just because pass phrase was present does not mean this was an issue with the pass phrase. It could be a wrong key, for example.

Reply:
Also it looks like the return errors should have one more item to be clearer, which would be `WRONG_PASS_PHRASE`

Reply:
There is no absolutely precise way to tell what has happened. We only can guess. So the guess is if we could not restore session key for some reason and there is passphrase, the cause is likely wrong password, otherwise wrong key. This is similar to what Typescript does.

Reply:
As far as I can tell, TypeScript we know with certainty which situation occurred.

If we cannot tell here, then we need to improve PGPainless so that it throws clearer errors.

But for pass phrase - and this also applies to https://github.com/FlowCrypt/flowcrypt-android/pull/1216#discussion_r628691330 - until we include that functionality in PGPainless, we could simply decrypt each key ahead of time before passing into PGPainless.